File: csharp/examples/Toolnexus.Examples/Acp.cs
using Toolnexus.Acp;

namespace Toolnexus.Examples;

/// <summary>
/// ACP (Agent Client Protocol) as the model behind the unified client (issue #96, ADR 0031):
/// a local coding-agent CLI — <c>devin acp</c> or <c>opencode acp</c> — is spawned ONCE, and
/// every turn below reuses that same warm session instead of paying process-startup cost again.
///
/// <para><b>Honesty check, read before you time anything:</b> whatever speedup turn 2/3 show
/// over turn 1 here is the CLI's own process-startup cost amortised across turns of ONE warm
/// session — it is NOT a protocol-level speedup, and ACP itself adds framing overhead, not
/// less. This example needs the agent CLI installed and already authenticated on this machine
/// (real <c>devin</c> or <c>opencode</c> credentials), so it is not hermetic and CI does not
/// run it — see <c>AcpClientTests</c> for the hermetic, fake-server-backed coverage of the
/// same client.</para>
///
/// <para>Agent CLI is selectable, not hardcoded:
/// <code>
///   dotnet run -- acp                                          # default: devin acp
///   ACP_AGENT_CMD=opencode ACP_AGENT_ARGS=acp dotnet run -- acp # or: opencode acp
///   dotnet run -- acp opencode acp                              # or pass the whole command
/// </code>
/// </para>
/// </summary>
internal static class Acp
{
    public static async Task<int> Run(string[] extraArgs)
    {
        var (command, arguments) = ResolveCommand(extraArgs);
        Console.WriteLine($"ACP agent command: {command} {string.Join(' ', arguments)}");

        AcpClient client;
        try
        {
            client = await AcpClient.StartAsync(command, arguments, cwd: Directory.GetCurrentDirectory());
        }
        catch (Exception e)
        {
            Console.WriteLine($"(could not spawn ACP agent '{command} {string.Join(' ', arguments)}' — "
                + $"is it installed and on PATH? {e.Message})");
            return 0;
        }

        try
        {
            var clock = NativeTool.Of("clock", "Return the current UTC time (ISO-8601).",
                new Dictionary<string, object?> { ["type"] = "object", ["properties"] = new Dictionary<string, object?>() },
                _ => DateTime.UtcNow.ToString("O"));

            await using var tk = await Toolkit.CreateAsync(new Toolkit.Options
            {
                ExtraTools = new List<ITool> { clock },
            });

            var agent = InProcess.CreateClient(new InProcess.Options
            {
                Model = "acp-agent",
                Generate = client.Generate,
                SystemPrompt = "You are a precise agent. Use the clock tool when asked for the time.",
            });

            string[] turns =
            {
                "What time is it right now? Use the clock tool.",
                "Thanks. Now just say 'ready' — no tool needed.",
            };

            foreach (var (turn, i) in turns.Select((t, i) => (t, i)))
            {
                var sw = System.Diagnostics.Stopwatch.StartNew();
                var res = await agent.RunAsync(turn, tk);
                sw.Stop();
                Console.WriteLine($"turn {i + 1} ({sw.ElapsedMilliseconds}ms), session={client.SessionId}: {Clip(res.Text, 120)}");
            }
        }
        finally
        {
            client.Dispose();
        }

        Console.WriteLine("\nOK ACP-backed client completed its turns on one warm session");
        return 0;
    }

    /// <summary>
    /// Resolve the ACP agent's command line: extra CLI args win if given (a whole command, e.g.
    /// <c>opencode acp</c>), else <c>ACP_AGENT_CMD</c>/<c>ACP_AGENT_ARGS</c> env vars, else the
    /// default <c>devin acp</c>. <c>opencode acp</c> works exactly the same way — swap the
    /// command, nothing else in this example changes.
    /// </summary>
    internal static (string Command, IReadOnlyList<string> Arguments) ResolveCommand(string[] extraArgs)
    {
        if (extraArgs.Length > 0)
        {
            return (extraArgs[0], extraArgs.Skip(1).ToList());
        }
        var cmd = Environment.GetEnvironmentVariable("ACP_AGENT_CMD");
        if (string.IsNullOrEmpty(cmd)) cmd = "devin";
        var argsEnv = Environment.GetEnvironmentVariable("ACP_AGENT_ARGS");
        if (string.IsNullOrEmpty(argsEnv)) argsEnv = "acp";
        var args = argsEnv.Split(' ', StringSplitOptions.RemoveEmptyEntries).ToList();
        return (cmd, args);
    }

    private static string Clip(string? s, int n)
    {
        if (string.IsNullOrEmpty(s)) return "";
        var oneLine = s.Replace("\n", " ");
        return oneLine.Length > n ? oneLine[..n] : oneLine;
    }
}

File: java/src/main/java/io/github/muthuishere/toolnexus/examples/Acp.java
package io.github.muthuishere.toolnexus.examples;

import io.github.muthuishere.toolnexus.InProcess;
import io.github.muthuishere.toolnexus.LlmClient;
import io.github.muthuishere.toolnexus.Tool;
import io.github.muthuishere.toolnexus.Toolkit;
import io.github.muthuishere.toolnexus.Tools;
import io.github.muthuishere.toolnexus.acp.AcpClient;
import io.github.muthuishere.toolnexus.annotations.Param;
import io.github.muthuishere.toolnexus.annotations.ToolMethod;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * ACP (Agent Client Protocol) as the model behind the unified client (issue #96, ADR 0025):
 * a local coding-agent CLI — {@code devin acp} or {@code opencode acp} — is spawned ONCE, and
 * every turn below reuses that same warm session instead of paying process-startup cost again.
 *
 * <p><b>Honesty check, read before you time anything:</b> whatever speedup turn 2/3 show over
 * turn 1 here is the CLI's own process-startup cost amortised across turns of ONE warm session —
 * it is NOT a protocol-level speedup, and ACP itself adds framing overhead, not less. This
 * example needs the agent CLI installed and already authenticated on this machine (real `devin`
 * or `opencode` credentials), so it is not hermetic and CI does not run it — see
 * {@code AcpClientTest} for the hermetic, fake-server-backed coverage of the same client.
 *
 * <p>Agent CLI is selectable, not hardcoded:
 * <pre>
 *   gradle runAcp                                   # default: devin acp
 *   ACP_AGENT_CMD=opencode ACP_AGENT_ARGS=acp gradle runAcp   # or: opencode acp
 *   gradle runAcp --args="opencode acp"             # or pass the whole command as args
 * </pre>
 */
public final class Acp {

    /** A plain object with an annotated tool method — proves the tool-calling loop is
     * unchanged: the ACP agent calls this exactly like it would call any MCP/HTTP/native tool. */
    public static final class ClockTools {
        @ToolMethod(name = "clock", description = "Return the current UTC time (ISO-8601).")
        public String clock() {
            return Instant.now().toString();
        }
    }

    public static void main(String[] args) throws Exception {
        List<String> command = resolveCommand(args);
        System.out.println("ACP agent command: " + String.join(" ", command));

        AcpClient client;
        try {
            client = AcpClient.start(command);
        } catch (Exception e) {
            System.out.println("(could not spawn ACP agent " + command + " — is it installed and on PATH? "
                    + e.getMessage() + ")");
            System.exit(0);
            return;
        }

        try (Toolkit tk = Toolkit.create(new Toolkit.Options())) {
            for (Tool t : Tools.fromObject(new ClockTools())) {
                tk.register(t);
            }

            LlmClient agent = InProcess.createClient(new InProcess.Options()
                    .model("acp-agent")
                    .generate(client)
                    .systemPrompt("You are a precise agent. Use the clock tool when asked for the time."));

            String[] turns = {
                    "What time is it right now? Use the clock tool.",
                    "Thanks. Now just say 'ready' — no tool needed.",
            };

            for (int i = 0; i < turns.length; i++) {
                long start = System.nanoTime();
                LlmClient.RunResult res = agent.run(turns[i], tk);
                long elapsedMs = (System.nanoTime() - start) / 1_000_000;
                System.out.println("turn " + (i + 1) + " (" + elapsedMs + "ms), session=" + client.sessionId()
                        + ": " + clip(res.text, 120));
            }
        } finally {
            client.close();
        }

        System.out.println("\nOK ACP-backed client completed its turns on one warm session");
    }

    /**
     * Resolve the ACP agent's command line: CLI args win if given (a whole command, e.g.
     * {@code opencode acp}), else {@code ACP_AGENT_CMD}/{@code ACP_AGENT_ARGS} env vars, else
     * the default {@code devin acp}. {@code opencode acp} works exactly the same way — swap
     * the command, nothing else in this example changes.
     */
    static List<String> resolveCommand(String[] args) {
        if (args != null && args.length > 0) {
            return new ArrayList<>(Arrays.asList(args));
        }
        String cmd = System.getenv("ACP_AGENT_CMD");
        if (cmd == null || cmd.isEmpty()) cmd = "devin";
        String argsEnv = System.getenv("ACP_AGENT_ARGS");
        if (argsEnv == null || argsEnv.isEmpty()) argsEnv = "acp";
        List<String> full = new ArrayList<>();
        full.add(cmd);
        for (String a : argsEnv.trim().split("\\s+")) {
            if (!a.isEmpty()) full.add(a);
        }
        return full;
    }

    private static String clip(String s, int n) {
        if (s == null) return "";
        String oneLine = s.replace("\n", " ");
        return oneLine.length() > n ? oneLine.substring(0, n) : oneLine;
    }
}
